### PR TITLE
Multiple validation rule targets

### PR DESCRIPTION
**Pull request by dmytro-grankin**
This PR brings support for multiple validation rule targets.

A definition of a validation rule with multiple  targets will look like:
```proto
message Rule {
    option (validation_of) = "foo.bar.UpdateDescription.change,"
                             "foo.bar.DescriptionUpdated.change";

    // A definition of constraints is omitted.
}
```

Other changes:
- `OptionParser` was introduced;
- classes related to validation rules were moved to `io.spine.validate.rules` (internal package).

### CODE REVIEW

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: base/src/main/java/io/spine/option/RawListParser.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * @param <R> the type of an element to be returned after parsing
+ * @author Dmytro Grankin
+ */
+public abstract class RawListParser<O extends ExtendableMessage, D extends GeneratedMessageV3, R>
+        implements OptionParser<D, R> {
+
+    /**
+     * The separator for the list values.
+     *
+     * <p>This class defines a concrete separator, because it used across
```
Comment (armiol) [inline base/src/main/java/io/spine/option/RawListParser.java]: `<p>This separator is used across the framework`. And drop the rest.

File: base/src/main/java/io/spine/option/RawListParser.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+     * @param value the option value to split
+     * @return the separated parts of the value
+     */
+    protected static Collection<String> splitOptionValue(CharSequence value) {
+        final String[] parts = PATTERN_VALUES_SEPARATOR.split(value);
+        return ImmutableList.copyOf(parts);
+    }
+
+    /**
+     * Obtains the number of the option to parse.
```
Comment (armiol) [inline base/src/main/java/io/spine/option/RawListParser.java]: I think, that's index? Let's discuss the name of the value.

File: base/src/main/java/io/spine/option/RawListParser.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+     *
+     * @return the option number
+     */
+    protected int getOptionNumber() {
+        return optionNumber;
+    }
+
+    public static String getValuesSeparator() {
+        return VALUES_SEPARATOR;
```
Comment (armiol) [inline base/src/main/java/io/spine/option/RawListParser.java]: Also, that's a `VALUE_SEPARATOR` and `getValueSeparator()`.

File: base/src/main/java/io/spine/option/TypeNamesParser.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import static io.spine.option.UnknownOptions.getUnknownOptionValue;
+import static java.util.Collections.emptyList;
+import static java.util.regex.Pattern.compile;
+
+/**
+ * A parser for the options, that represent {@linkplain RawListParser raw list} of type names.
+ *
+ * @author Dmytro Grankin
+ */
+public class TypeNamesParser extends RawListParser<MessageOptions, DescriptorProto, TypeName> {
```
Comment (armiol) [inline base/src/main/java/io/spine/option/TypeNamesParser.java]: This whole class is reported to be untested.

File: base/src/main/java/io/spine/option/ValidationTargetsParser.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ */
+public class ValidationTargetsParser extends RawListParser<MessageOptions, DescriptorProto, String> {
+
+    private ValidationTargetsParser() {
+        super(validationOf);
+    }
+
+    @Override
+    public Collection<String> parse(DescriptorProto descriptor) {
+        final String value = getUnknownOptionValue(descriptor, getOptionNumber());
```
Comment (armiol) [inline base/src/main/java/io/spine/option/ValidationTargetsParser.java]: This method is reported to be untested.

File: base/src/main/java/io/spine/option/ValidationTargetsParser.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        if (value == null) {
+            return emptyList();
+        }
+
+        return parse(value);
+    }
+
+    @Override
+    public Collection<String> parse(String optionValue) {
+        checkArgument(!isNullOrEmpty(optionValue));
```
Comment (armiol) [inline base/src/main/java/io/spine/option/ValidationTargetsParser.java]: Only one of two branches is tested.

File: base/src/main/java/io/spine/validate/rules/ValidationRule.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.util.Collection;
+
+import static com.google.protobuf.Descriptors.FieldDescriptor.JavaType.MESSAGE;
+import static io.spine.util.Exceptions.newIllegalStateException;
+
+/**
+ * A {@linkplain io.spine.option.OptionsProto#validationOf validation rule}.
+ *
+ * <p>Aggregates descriptors related to the validation rule.
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRule.java]: `... to a particular validation rule.`

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import com.google.common.collect.ImmutableSet;
+import io.spine.option.ValidationTargetsParser;
+
+import java.util.Collection;
+import java.util.Properties;
+
+import static io.spine.util.PropertyFiles.loadAllProperties;
+
+/**
+ * Utilities for obtaining {@linkplain ValidationRule validation rules} known for the application.
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: `to the application`.

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * <p>During initialization of this class, definitions of validation rules are
+ * {@linkplain Builder#put(Properties) validated}. If an invalid validation rule was found,
+ * a runtime exception will be raised.
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: `thrown.`

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * @author Dmytro Grankin
+ */
+public class ValidationRules {
+
+    /**
+     * A name of the file, which contains validation rules and their target fields paths.
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: `target field paths`

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    private ValidationRules() {
+        // Prevent instantiation of this class.
+    }
+
+    /**
+     * Obtains validation rules known to the application.
+     *
+     * @return the immutable collection of validation rules
+     */
+    @SuppressWarnings("ReturnOfCollectionOrArrayField") // It's ok to return
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: `OK`

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        return rules;
+    }
+
+    public static String getValRulesPropsFileName() {
+        return VAL_RULES_PROPS_FILE_NAME;
+    }
+
+    /**
+     * {@code Builder} assembles the validation rules from the
+     * {@linkplain #VAL_RULES_PROPS_FILE_NAME properties} files.
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: Probably, `file` (singular). Or, if plural, please append something about classpath scan — most likely that would be a part of the string constant description.

File: base/src/main/java/io/spine/validate/rules/package-info.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Classes for working with validation rules.
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/package-info.java]: `This package defines classes and interfaces for working with {@code Message} validation rules.`

File: base/src/main/java/io/spine/option/RawListParser.java (hunk 1 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/RawListParser.java]: Done.

File: base/src/main/java/io/spine/option/RawListParser.java (hunk 3 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/RawListParser.java]: Done.

File: base/src/main/java/io/spine/option/TypeNamesParser.java (hunk 4 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/TypeNamesParser.java]: Done.

File: base/src/main/java/io/spine/option/ValidationTargetsParser.java (hunk 5 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/ValidationTargetsParser.java]: Done.

File: base/src/main/java/io/spine/option/ValidationTargetsParser.java (hunk 6 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/ValidationTargetsParser.java]: Done.

File: base/src/main/java/io/spine/validate/rules/ValidationRule.java (hunk 7 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRule.java]: Done.

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java (hunk 8 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: Done.

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java (hunk 9 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: Done.

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java (hunk 10 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: Done.

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java (hunk 11 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: Done.

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java (hunk 12 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: Done.

File: base/src/main/java/io/spine/validate/rules/package-info.java (hunk 13 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/package-info.java]: Done.

File: base/src/main/java/io/spine/option/RawListParser.java (hunk 2 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/RawListParser.java]: Done.

**Comment by dmytro-grankin** [comment]
@armiol PTAL. Let's increase the version to keep the difference between versions of the latest `model-compiler` and `model-compiler`, which is used during `base` build.

File: base/src/main/java/io/spine/option/RawListParser.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Override
+    public Collection<R> parse(D descriptor) {
+        final String optionValue = getOptionValue(descriptor, optionNumber);
+        if (optionValue == null) {
+            return emptyList();
+        }
+
+        return parse(optionValue);
```
Comment (armiol) [inline base/src/main/java/io/spine/option/RawListParser.java]: This line is reported to be untested.

File: base/src/main/java/io/spine/option/TypeNamesParser.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+    public TypeNamesParser(GeneratedExtension<MessageOptions, String> option,
+                           String packagePrefix) {
+        super(option);
+        this.packagePrefix = checkNotNull(packagePrefix);
+    }
+
+    @Override
+    protected String getOptionValue(DescriptorProto descriptor, int optionNumber) {
+        return getUnknownOptionValue(descriptor, optionNumber);
```
Comment (armiol) [inline base/src/main/java/io/spine/option/TypeNamesParser.java]: This line is reported to be untested.

File: base/src/main/java/io/spine/option/ValidationTargetsParser.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public class ValidationTargetsParser extends RawListParser<MessageOptions, DescriptorProto, String> {
+
+    private ValidationTargetsParser() {
+        super(validationOf);
+    }
+
+    @Override
+    protected String getOptionValue(DescriptorProto descriptor, int optionNumber) {
+        return getUnknownOptionValue(descriptor, optionNumber);
```
Comment (armiol) [inline base/src/main/java/io/spine/option/ValidationTargetsParser.java]: Same as above.

File: base/src/main/java/io/spine/validate/rules/ValidationRule.java [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.util.Collection;
+
+import static com.google.protobuf.Descriptors.FieldDescriptor.JavaType.MESSAGE;
+import static io.spine.util.Exceptions.newIllegalStateException;
+
+/**
+ * A {@linkplain io.spine.option.OptionsProto#validationOf validation rule}.
+ *
+ * <p>Aggregates descriptors related to a particular validation rule.
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRule.java]: Let's not use `to aggregate` verb in all cases except when talking about `Aggregate`s.

File: base/src/main/java/io/spine/validate/rules/ValidationRule.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+     * The descriptor for the validation rule message.
+     */
+    private final Descriptor descriptor;
+
+    /**
+     * Descriptors for the target fields of the validation rule.
+     */
+    private final Collection<FieldDescriptor> targets;
+
+    ValidationRule(String typeName, Iterable<String> targetPaths) {
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRule.java]: This is a suspicious ctor. Let's discuss its usage.

File: base/src/main/proto/spine/options.proto [hunk 19]
```diff
@@ -302,12 +301,14 @@ extend google.protobuf.MessageOptions {
     // If a specified options is already defined in a target message,
     // then options from the validation rule will be used.
     //
-    // Validation rules are applicable only for field declarations of a `Message` type.
-    //
-    // Validation rule target should be marked as `valid`. Otherwise a rule will be ignored.
+    // Validation rule limitations:
+    // - applicable only for field declarations of a `Message` type;
```
Comment (armiol) [inline base/src/main/proto/spine/options.proto]: In this case `-applicable only to field declarations ...`.

File: base/src/test/proto/spine/test/validate/msg/messages.proto [hunk 20]
```diff
@@ -242,14 +242,25 @@ message MessageWithMapDoubleField {
 
 // Messages for testing validation rules.
 
-message ValidationRule {
-    option (validation_of) = "spine.test.validate.msg.ValidationRuleTarget.can_be_valid";
+message AValidationRule {
+    option (validation_of) = "spine.test.validate.msg.FirstRuleTarget.can_be_valid,"
+                             "spine.test.validate.msg.SecondRuleTarget.can_be_valid";
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/messages.proto]: What if there is a space before `spine.test.validate.msg.SecondRuleTarget.can_be_valid`? E.g.:
```
"    spine.test.validate.msg.SecondRuleTarget.can_be_valid"
```
The feature should work fine (you should've stripped the whitespace). However I don't see a test covering this case.

File: base/src/main/java/io/spine/validate/rules/ValidationRule.java (hunk 17 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRule.java]: Done.

File: base/src/main/proto/spine/options.proto (hunk 19 above)
Comment (dmytro-grankin) [inline base/src/main/proto/spine/options.proto]: Done.

File: base/src/main/java/io/spine/option/RawListParser.java (hunk 14 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/RawListParser.java]: Done.

File: base/src/main/java/io/spine/option/TypeNamesParser.java (hunk 15 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/TypeNamesParser.java]: Done.

File: base/src/main/java/io/spine/option/ValidationTargetsParser.java (hunk 16 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/option/ValidationTargetsParser.java]: Done.

File: base/src/test/proto/spine/test/validate/msg/messages.proto (hunk 20 above)
Comment (dmytro-grankin) [inline base/src/test/proto/spine/test/validate/msg/messages.proto]: There is the test for the case.

File: base/src/main/java/io/spine/validate/rules/ValidationRule.java (hunk 18 above)
Comment (dmytro-grankin) [inline base/src/main/java/io/spine/validate/rules/ValidationRule.java]: Done.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: base/src/main/java/io/spine/validate/rules/ValidationRules.java (hunk 12 above)
Comment (armiol) [inline base/src/main/java/io/spine/validate/rules/ValidationRules.java]: What I meant was that you should've explained that there were multiple files in the classpath; all of those were taken into an account; if there were some duplicates in keys — [then what].  Etc.

**Pull request merged**